### PR TITLE
Fix double @RequestMapping with empty value

### PR DESCRIPTION
**Pull request by nekkiy**

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@nekkiy Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-openfeign&pullRequestId=245)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-openfeign&pullRequestId=245) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@nekkiy Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-openfeign&pullRequestId=245)!

**Comment by nekkiy** [comment]
@OlgaMaciaszek conflicts resolved

**Comment by OlgaMaciaszek** [comment]
Thanks @nekkiy - will have a look at it today.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java [hunk 1]
```diff
@@ -569,13 +570,50 @@ public void testAddingTemplatedParameterWithTheSameKey()
 				"{Accept}");
 	}
 
+	@Test
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: Please convert this test into more tests that are smaller and more readable. Each test should have a visually divided `given` `when` and `then` sections - only 1 of each section in one test (without multiple `when-then` blocks in 1 test as it's now).

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java [hunk 2]
```diff
@@ -167,15 +167,14 @@ protected void processAnnotationOnClass(MethodMetadata data, Class<?> clz) {
 		if (clz.getInterfaces().length == 0) {
 			RequestMapping classAnnotation = findMergedAnnotation(clz,
 					RequestMapping.class);
-			if (classAnnotation != null) {
+			if (classAnnotation != null && classAnnotation.value().length > 0) {
 				// Prepend path from class annotation if specified
-				if (classAnnotation.value().length > 0) {
-					String pathValue = emptyToNull(classAnnotation.value()[0]);
+				String pathValue = emptyToNull(classAnnotation.value()[0]);
+				if (pathValue != null) {
```
[the remark this answers is not in the document]
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java]: :muscle:

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java [hunk 3]
```diff
@@ -569,13 +570,50 @@ public void testAddingTemplatedParameterWithTheSameKey()
...
+			assertThat(dataRoot.template().url()).isEqualTo("/");
+		}
+
+	}
+
 	@Test
 	public void testMultipleRequestPartAnnotations() throws NoSuchMethodException {
 		Method method = TestTemplate_RequestPart.class.getDeclaredMethod(
-				"requestWithMultipleParts", MultipartFile.class, String.class);
+			"requestWithMultipleParts", MultipartFile.class, String.class);
```
[the remark this answers is not in the document]
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: fix

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java [hunk 4]
```diff
@@ -569,13 +570,50 @@ public void testAddingTemplatedParameterWithTheSameKey()
...
 
 		MethodMetadata data = contract
-				.parseAndValidateMetadata(method.getDeclaringClass(), method);
+			.parseAndValidateMetadata(method.getDeclaringClass(), method);
```
[the remark this answers is not in the document]
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: fix

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java [hunk 5]
```diff
@@ -69,6 +69,7 @@
  * @author Halvdan Hoem Grelland
  * @author Aram Peres
  * @author Aaron Whiteside
+ * @author nekkiy
```
[the remark this answers is not in the document]
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: fix

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java [hunk 6]
```diff
@@ -229,11 +228,10 @@ protected void processAnnotationOnMethod(MethodMetadata data,
 			String pathValue = emptyToNull(methodMapping.value()[0]);
 			if (pathValue != null) {
 				pathValue = resolve(pathValue);
-				// Append path from @RequestMapping if value is present on method
-				if (!pathValue.startsWith("/") && !data.template().path().endsWith("/")) {
-					pathValue = "/" + pathValue;
+				if (!pathValue.equals("/")) {
+					// Automatic add '/' if it did not present in pathValue
```
[the remark this answers is not in the document]
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java]: Comment was dropped.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java (hunk 1 above)
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: Unfortunately, JUnit 4.x doesn't support parameters for separately test or group of tests inside one test class/ Junit 4.x also doesn't support nested tests. So I think, that separating to more tests can lead to splitting of logic. But, if it really need I can do it.

**Comment by nekkiy** [comment]
Hi, @OlgaMaciaszek Thank you for the review. Empty paths to `@RequestMapping` may is frequent situation, because `@RequestMapping` also have `method` parameter. I can’t say it’s good practice) But my case it is using of code generator (https://github.com/OpenAPITools/openapi-generator). Next code is generated for `kotlin-spring` profile:
```kotlin
@RequestMapping("\${api.base-path:}")
interface MyApi {
    @RequestMapping(
            value = ["/myMethod"],
            produces = ["application/json"], 
            method = [RequestMethod.GET])
    fun myMethod() : ResponseEntity<String>
}

```
If `api.base-path` is empty then I'm taken result path with double slash such as `//myMethod`.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java (hunk 1 above)
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: It does: https://github.com/junit-team/junit4/wiki/Parameterized-tests.
In any case, please rather divide it in order to make it more readable. You can have various tests for the various scenarios you're testing.

Comment (nekkiy) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: Unfortunately, native JUnit 4.x mechanism for parameterizing may clog nearby tests. I split test using https://github.com/Pragmatists/JUnitParams. Will it fit?

Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java]: Looks good. Thanks.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java [hunk 7]
```diff
@@ -229,11 +229,9 @@ protected void processAnnotationOnMethod(MethodMetadata data,
 			String pathValue = emptyToNull(methodMapping.value()[0]);
 			if (pathValue != null) {
 				pathValue = resolve(pathValue);
-				// Append path from @RequestMapping if value is present on method
-				if (!pathValue.startsWith("/") && !data.template().path().endsWith("/")) {
-					pathValue = "/" + pathValue;
+				if (!pathValue.equals("/")) {
```
Comment (adriancole) [inline spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java:232]: this actually breaks things as it eats "/" resulting in a path "" even if the user specifies "/" in their annotation. Can this be handled differently?

**Comment by adriancole** [comment]
this revealed a bug in sleuth where "" was the only path we skipped by default for clients! https://github.com/spring-cloud/spring-cloud-sleuth/pull/1598

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java (hunk 7 above)
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java:232]: yes, that's right; will fix it.

**Comment by OlgaMaciaszek** [comment]
ok. will revert it for now, to the code without it in today's `Hoxton` release. And then we could think of a better fix.

**Comment by OlgaMaciaszek** [comment]
@nekkiy I had to revert it [here](https://github.com/spring-cloud/spring-cloud-openfeign/commit/14929cbfee02580a8b0684bb665b7b335f957aa6) due to [this issue](https://github.com/spring-cloud/spring-cloud-openfeign/issues/386). Let me know if you would like to work on an alternative implementation.

**Comment by nekkiy** [comment]
I could work on this after 15 August.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java (hunk 7 above)
Comment (nekkiy) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java:232]: @adriancole, @OlgaMaciaszek  Sorry for delay. 
> this actually breaks things as it eats "/" resulting in a path "" even if the user specifies "/" in their annotation.

No. See source code from [RequestTemplate](https://github.com/OpenFeign/feign/blob/10.10.1/core/src/main/java/feign/RequestTemplate.java#L527)
RequestTemplate
I wrote [new test in my branch to demonstrate it](https://github.com/nekkiy/spring-cloud-openfeign/blob/fixEmptyRequestMapping_v2/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/support/SpringMvcContractTests.java#L606).  

@adriancole take me more details to I could resolve it. Best way is you take code of class with broken part.

Best regard.

**Comment by shark300** [comment]
Dear @nekkiy,
We have a same issue with that, please let me know what is the status of this issue?

**Pull request merged**